### PR TITLE
Implement API for /questions route

### LINKED ISSUE #36: API Creation
I can think of the following
- GET Quote and OPTIONS
-  UPDATE Score

Please let us know your suggestions here.

### PR DESCRIPTION
**Pull request by ashwani99**
- Added route for `/questions`
- Added controller for `GET /questions`
- limit query param should be used to get `n` number of quotes as JSON

Fixes #36

### CODE REVIEW

File: server/api/controllers/question.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        });
+    }
+
+    Question.count()
+        .exec()
+        .then(count => {
+            const random = Math.floor(Math.random() * count);
+            return Question.find()
+                .skip(random)
+                .limit(limit)
```
Comment (skyshader) [inline server/api/controllers/question.js]: What do you want to accomplish by using skip and limit?

File: server/api/controllers/question.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+const { Question } = require("../models/question");
+
+exports.get_questions = (req, res) => {
```
Comment (skyshader) [inline server/api/controllers/question.js]: Keep the names short and sweet. As the method is already inside question controller, a simple name as `get` would be enough.

File: server/api/routes/index.js [hunk 3]
```diff
@@ -2,6 +2,9 @@ const express = require('express');
 const router = express.Router();
 
 const userRoutes = require('./user');
-router.use('/user',userRoutes);
+const apiRoute = require('./question');
```
Comment (skyshader) [inline server/api/routes/index.js]: rename to questionRoutes?

File: server/api/controllers/question.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    Question.count()
+        .exec()
+        .then(count => {
+            const random = Math.floor(Math.random() * count);
+            return Question.find()
+                .skip(random)
+                .limit(limit)
+                .exec()
+        })
+        .then(quotes_json => {
```
Comment (skyshader) [inline server/api/controllers/question.js]: let's rename it as `quotes`

File: server/api/routes/question.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+const express = require('express');
+const router = express.Router();
+const questionController = require('../controllers/question');
+
+router.get('/questions', questionController.get_questions);
```
Comment (skyshader) [inline server/api/routes/question.js]: `questionController.get`

File: server/app.js [hunk 6]
```diff
@@ -2,6 +2,8 @@ const express = require("express");
 const morgan = require("morgan");
 const cors = require("cors");
 const bodyParser = require("body-parser");
+const fs = require('fs');
+const { Question } = require("./api/models/question");
```
Comment (skyshader) [inline server/app.js]: aren't we just exporting the Question model from the file? Do we really need `{}`

File: server/app.js [hunk 7]
```diff
@@ -30,4 +32,12 @@ app.use((err, req, res) => {
   });
 });
 
+// initialize database
+fs.readFile('seed/quotes.json', (err, data) => {
```
Comment (skyshader) [inline server/app.js]: lets move it to seed directory and call the file `seeder.js`
also let's add a `npm` script to seed the database `npm run seed`

make sure you clear the database before starting to insert

**Review by skyshader — CHANGES_REQUESTED** [review]

File: server/api/controllers/question.js [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    if (!limit) {
+        res.status(404).json({
+            error: 'limit is not provided'
+        });
+    }
+
+    Question.count()
+        .exec()
+        .then(count => {
+            const random = Math.floor(Math.random() * count);
```
[the remark this answers is not in the document]
Comment (ashwani99) [reply server/api/controllers/question.js]: No, did you test the API?

File: server/api/controllers/question.js (hunk 1 above)
Comment (skyshader) [reply server/api/controllers/question.js]: Interesting. But after skipping a certain number of questions we would not have a mix of random questions. Let's say for random numbers 1, 2, 3, 4, 5, 6 and following the questions sequence will be same (For 1: 2, 3, 4, 5, 6 ::::  2: 3, 4, 5, 6, 7 ) see how questions overlap.

Why don't we try getting the `limit` number of unique random numbers within `count` and then return those set of questions?

File: server/app.js (hunk 6 above)
Comment (ashwani99) [reply server/app.js]: Yes we need it. Please see this https://github.com/UdacityMobileWebScholarship/guess-quote/blob/49181c9fc821e9e8a69c974911c5919d9b0c9e9d/server/api/models/question.js#L38

File: server/app.js (hunk 7 above)
Comment (ashwani99) [reply server/app.js]: @skyshader What about I create a function in `seeder.js` and call that from `server.js` or `app.js`?

File: server/app.js (hunk 6 above)
Comment (skyshader) [reply server/app.js]: Ok then let's export it directly. Without the braces.

File: server/app.js (hunk 7 above)
Comment (skyshader) [reply server/app.js]: Nope, app should be independent of seeder. A seeder should be involved independently from the app.js.

File: server/api/controllers/question.js [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+      error: { message: "limit is not provided" }
+    });
+  }
+
+  let options = { limit: limit };
+  Question.findRandom({}, {}, options, (err, quotes) => {
+    if (!err) {
+      res.send(quotes);
+    } else {
+      res.status(500).json({
```
Comment (skyshader) [inline server/api/controllers/question.js]: Let's reverse the check and return the exact error if there is error.

Later if there are no quotes return 200 status with empty quotes array.

And finally return the quotes.

File: server/api/controllers/question.js [hunk 10]
```diff
@@ -0,0 +1,22 @@
+const Question = require("../models/question");
+
+exports.get = (req, res) => {
+  const limit = Number(req.query.limit);
+
+  if (!limit) {
+    res.status(404).json({
```
Comment (skyshader) [inline server/api/controllers/question.js]: 404 doesn't make sense here
Let's use 400 which means Bad Request as this is not what we accept from the client.

File: server/api/controllers/question.js (hunk 9 above)
Comment (ashwani99) [reply server/api/controllers/question.js]: @skyshader I don't think its a good idea to return the exact server error to the client. Instead we can  throw the error on the server machine and send something like "Server error occurred" with a 500

**Comment by skyshader** [comment]
Awesome. LGTM :+1:

**Comment by ashwani99** [comment]
Thanks @skyshader 
Wanna close #36 ?

**Pull request merged**